src/charm.py

```diff
@@ -0,0 +1,208 @@
+        self.framework.observe(self.on.remove, self._remove)
+        self.framework.observe(self.on.update_status, self._update_status)
+        # self.framework.observe(self.on.config_changed, self._reconcile)
+
+        self.x = 1
+        self.logger = logging.getLogger(__name__)
+        self.y = 1
+        # TODO: Fix file imports and move ./src/files back to ./files
+        self._manifests_file_root = None
+        self.manifest_file_root = "./src/files/manifests/"
+        self.image = OCIImageResource(self, "oci-image")
+
+    def _noop_pebble_ready(self, _):
+        self.logger.info("noop_pebble_ready fired")
+
+    def _install(self, event):
+        self.logger.info("Installing by instantiating Kubernetes objects")
+        self.unit.status = MaintenanceStatus("Instantiating Kubernetes objects")
+        _, manifests_str = self._render_manifests()
+
+        self.logger.info("Applying manifests")
+
+        completed_process = subprocess.run(
+            ["./kubectl", "apply", "-f-"],
+            input=manifests_str.encode("utf-8"),
+            stderr=subprocess.STDOUT,
+            stdout=subprocess.PIPE,
+        )
+        self.logger.info(
+            f"kubectl returned with code '{completed_process.returncode}'"
+            f" and output {completed_process.stdout}"
+        )
+
+        self.logger.info("Waiting for installed Kubernetes objects to be operational")
+        max_attempts = 20
+        for attempt in range(max_attempts):
+            self.logger.info(f"validation attempt {attempt}/{max_attempts}")
+            running = self._check_deployed_resources()
+
+            if running is True:
+                self.logger.info("Resources detected as running")
+                self.logger.info("Install successful")
+                self.unit.status = ActiveStatus()
+                return
+            else:
+                sleeptime = 10
+                self.logger.info(f"Sleeping {sleeptime}s")
+                time.sleep(sleeptime)
+        else:
+            self.unit.status = MaintenanceStatus(
+                "Some kubernetes resources missing/not ready"
+            )
+            return
+
+    def _update_status(self, event):
+        self.logger.info("Comparing current state to desired state")
+
+        running = self._check_deployed_resources()
+        if running is True:
+            self.logger.info("Resources are ok.  Unit in ActiveStatus")
+            self.unit.status = ActiveStatus()
+            return
+        else:
+            self.logger.info(
+                "Resources are missing.  Triggering install to reconcile resources"
+            )
+            self.unit.status = MaintenanceStatus(
+                "Missing kubernetes resources detected - reinstalling"
+            )
+            self._install(event)
+            return
+
+    def _remove(self, _):
+        """Remove charm"""
+        # TODO: How should we test this in our integration tests?
```

Comment:
pytest-operator will destroy the model when finishing its tests, but you can still at any time remove an application during the tests calling libjuju and it might be a nice test case to add.